parser/wasm_parse/mod.rs

```
@@ -15,50 +15,125 @@ pub struct SectionIndices {
+            index: index,
+            reader: wasmparser::CodeSectionReader::new(
+                &self.data[byte_range.start..byte_range.end],
+                byte_range.start,
+            )?,
+            byte_size: byte_range.end - start_offset,
+        })
+    }
+}
+
+impl<'a> Parse<'a> for ModuleReader<'a> {
     type ItemsExtra = ();
 
     fn parse_items(
         &mut self,
         items: &mut ir::ItemsBuilder,
         _extra: (),
     ) -> Result<(), traits::Error> {
-        let initial_offset = self.current_position();
         let mut sections: Vec<IndexedSection<'_>> = Vec::new();
-        let mut code_section: Option<IndexedSection<'_>> = None;
-        let mut function_section: Option<IndexedSection<'_>> = None;
+        let mut code_section: Option<CodeSection<'_>> = None;
+        let mut function_section: Option<FunctionSection<'_>> = None;
         let mut sizes: HashMap<usize, u32> = HashMap::new();
 
         // The function and code sections must be handled differently, so these
         // are not placed in the same `sections` array as the rest.
         let mut idx = 0;
-        while !self.eof() {
+        // @@@ revert to old style.
+        loop {
             let start = self.current_position();
-            let section = self.read()?;
-            let size = self.current_position() - start;
-            let indexed_section = IndexedSection(idx, section);
-            match indexed_section.1.code {
-                wasmparser::SectionCode::Code => code_section = Some(indexed_section),
-                wasmparser::SectionCode::Function => function_section = Some(indexed_section),
+            let at_eof = self.offset == self.data.len();
+            if at_eof {
+                break;
+            }
+            let (section, bytes_consumed) =
+                match self.parser.parse(&self.data[self.offset..], at_eof)? {
+                    wasmparser::Chunk::NeedMoreData { .. } => {
+                        panic!("@@@ nyi");
```

We should return an error here too.